README.txt
/* **** **** **** **** **** **** **** **** **** ***** *\
|                                                      |
|    W    O    R    D    A    M    E    N    T   ZERO  |
|                                                      |
\* **** **** **** **** **** **** **** **** **** ***** */

========================================================
Tugas Besar Alstrukdat 2014
Code Copyright: Group 3 K-2 IF 2013
========================================================

--- --- --- --- --- --- --- --- --- --- --- --- --- ---
1. STRUKTUR
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Struktur program dibagi menjadi 4 bagian (folder)

a. adt - berisi ADT yang digunakan dalam program

b. module - berisi program baru (yang menggunakan adt) 
            yang akan digabungkan menjadi game utuh

c. res - berisi data yang diberikan oleh spek seperti 
         jenis board dan kamus kata yang digunakan

d. saves - berisi data simpanan yang digenerate dari 
           user dan dapat dipanggil lagi jika dibutuhkan


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
2. ADT
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Folder ADT berisi ADT yang digunakan dalam rangkaian 
program.

Didalam folder ADT ini berisi folder-folder yang diberi
nama sesuai dengan jenis atau kegunaan dari ADT tersebut


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
3. MODULE
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Folder module adalah file-file bentukan dari ADT yang
akan digunakan dalam driver akhir pada permainan.

Penamaan disesuaikan dengan fungsi dari file tersebut.

a. save.c - menyimpan data dari masukkan pengguna, atau
            user-generated, kedalam file .txt di dalam
            folder saves

b. load.c - mengambil data dari folder saves sesuai deng
            an permintaan

c. word.c - berisi cara kerja permainan dan cara mengambil
            kata yang dibentuk oleh user ke dalam stack

d. dict.c - mengambil kamus, dan juga membentuknya. Berisi
	    juga fungsi hash yang digunakan serta pencarian
	    hash suatu kata

e. auth.c - berisi fungsi untuk memvalidasi suatu login dan
	    register

NOTE: 
Gunakan global.h untuk menambah variabel global, 
fungsi baru atau ADT baru agar tertata rapi.

Sehingga ditiap file .c yang ada di module hanya meng-
include file global.c (kecuali driver, yang membutuhkan
file *.c lainnya)


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
4. RES
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Berisi file dari spek yang harus dan akan digunakan.


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
3. SAVES
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Terdiri dari 11 folder, 10 adalah folder berdasarkan 
board dan 1 adalah folder debug atau untuk test purpose.

Didalam folder tersebut terdapat folder lagi yakni:
highscore dan words.

Dimasing-masing folder, akan terdapat file txt yang nama
nya merupakan nama user yang memiliki data tersebut.
contoh: feryandi.txt

File txt didalam folder highscore akan mempunyai format
sebagai berikut:
<score> <dd> <mm> <yy> <hh> <MM>
<score> <dd> <mm> <yy> <hh> <MM>
.

File txt didalam folder words akan mempunyai format
sebagai berikut:
<kata yang pernah ditebak>
<kata yang pernah ditebak>
<kata yang pernah ditebak>
.

--- --- --- --- --- --- --- --- --- --- --- --- --- ---
4. UI
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Berisi cara kerja user-interface dan pembuatannya.


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
5. DISCLAIMER
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
Silahkan gunakan file ini sebagai salah satu bagian
dari changelog sehingga memudahkan pengkodingan bersama.


--- --- --- --- --- --- --- --- --- --- --- --- --- ---
6. RELEASE(S)
--- --- --- --- --- --- --- --- --- --- --- --- --- ---
ALPHA STAGE
v0.1 (7/11/2014)


--- --- --- --- --- END OF FILE --- --- --- --- ---


CHANGELOG.txt

CHANGELOG

---- ---- ---- ---- ---- ---- ---- ---- ---- ---- 
Format:
<tanggal dan waktu> - <pengubah>
<jenis ubahan>
<apa saja yang diubah>
---- ---- ---- ---- ---- ---- ---- ---- ---- ---- 

4/11/2014 23:24 - dummy
Tes Changelog
* Membuat file changelog.txt
* Membuat file readme.txt
* Tidur

5/11/2014 21:44 - feryandi
Login System
* Menambahkan file auth.c pada module
* Merubah driver.c
* Register
  Gunakan fungsi registerUser(<nama_user>), fungsi ini akan
  mengeluarkan boolean
* Login
  Gunakan fungsi loginUser(<nama_user>), fungsi ini akan
  mengeluarkan boolean
  
6/11/2014 23:15 - feryandi
Dictionary!
* import seluruh kata ke kamus menggunakan: 
			getDictionary(TabKamus (*T))
* cari tahu apakah ada kata tersebut dalam kamus menggunakan:
			isOnKamus (char word[18], TabKamus X) -> boolean

7/11/2014 23:07 - feryandi
Stack Overflow!
* Fixed: stack overflow, sekarang seluruh fungsi/prosedur di dict.c
	 tidak memerlukan inputan file kamus, karena variable kamus
	 telah dijadikan variabel global yang bisa dipakai kapanpun
	 dan dimanapun selama telah diinisiasi getDirectory(K);
* RELEASE: driverword.c - ver. alpha 0.1

9/11/2014 14:17 - rakarukri
BIG UPDATE
* Update UI (kalibrasi ulang koordinat)

* Sistem baru ( getch(void), gotoxy(void), clrscr(void), Hide/ShowCursor(void) dan warna text(di UI/interface.h))

* Sinkronisasi file lama (Windows) ke Linux

* Note : Hanya bisa dijalankan di Linux (driverword.c)

9/11/2014 20:14 - feryandi
Authentication
* Autentikasi sekarang dibutuhkan buat masuk ke game (menu register dan login sudah aktif)

9/11/2014 23:59 - rakarukri
UI - Authentication
* Autentikasi sudah disinkronisasi dengan UI
* Update UI login - register
* Revisi changelog = 10/11/2014 00:00 :P

10/11/2014 10:41 - rakarukri
UI
* Update UI login
* Implementasi pilihan board
* Error : setelah main belum bisa main lagi (coredump)

11/11/2014 10:56 - feryandi
Suggestion is ON!
* change in: word.c, global.c, adt/set, driverword.c, load.c

11/11/2014 21:40 - rakarukri
Wordament Zero release!
* Playable yeay!
* to do :
	1. kata yang sama jangan diproses lagi (bisa pake file eksternal baru) supaya score
	juga ngga nambah
	2. block pengisian username-register biar ngga jebol
	3. about,how to play, highscore
	4. PLAY THE GAME!
	


module/word.c
#include "global.h"
#include "dict.c"

/* Kamus Global */
extern boolean endGame;
int absis,ordinat,simpanabsis,simpanordinat, count,absisboard,ordinatboard;
boolean aktif, discovered[5][5];

void initBoard ()
{
    // Kamus Lokal
    int i, j;

	// Algoritma
    count = 0;
    absis = 28;
    ordinat = 7;
    for (i=1;i<=4;i++)
    {
        for (j=1;j<=4;j++)
    	{
            discovered[i][j] = false;
		}
	}
}

void PointerBoard(MATRIKS *M, Stack *S, char *kata, boolean *endK)
{
    // Kamus Lokal
    int i, j, score;
    char cc;

    // Algoritma
    gotoxy(absis,ordinat);


    switch(getch())
    {
    	case 'q':
    		{
				if((absis >= 36)&&(ordinat >= 11))
		    		{
						if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
						absis -= 8;
						ordinat -= 4;
						gotoxy(absis,ordinat);
						if(!aktif)
                    {
                    	PointerAktif();
                    }
						if (aktif == true)
						{
							SimbolAktif((*M), S);
		            	}
		            }
				break;
			}
		case 'w':
			{
				if (ordinat >= 11)
				{
					if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
					ordinat -= 4;
					gotoxy(absis,ordinat);
					if(!aktif)
                    {
                    	PointerAktif();
                    }
					if (aktif == true)
					{
						SimbolAktif((*M), S);
					}
				}
				break;
			}
		case 'e':
			{
				if ((absis <= 44)&&(ordinat >= 10))
				{
					if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
					absis += 8;
					ordinat -= 4;
					gotoxy(absis,ordinat);
					if(!aktif)
                    {
                    	PointerAktif();
                    }
					if (aktif == true)
					{
						SimbolAktif((*M), S);
					}
				}
				break;
			}
        case 'a':
            {
                if (absis >= 36)
                {
                    if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
                    absis -= 8;
                    gotoxy(absis,ordinat);
                    if(!aktif)
                    {
                    	PointerAktif();
                    }
                    if (aktif == true)
                    {
                        SimbolAktif((*M), S);
                	}
                }
                break;
            }
        case 'd':
            {
                if (absis <= 44)
                {
                    if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
                    absis += 8;
                    gotoxy(absis,ordinat);
                    if(!aktif)
                    {
                    	PointerAktif();
                    }
                    if (aktif == true)
                    {
                        SimbolAktif((*M), S);
                	}
                }
                break;
            }
        case 'z':
            {
                if ((absis >= 36)&&(ordinat <= 15))
                {
                    if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
                    absis -= 8;
                    ordinat += 4;
                    gotoxy(absis,ordinat);
                    if(!aktif)
                    {
                    	PointerAktif();
                    }
                    if (aktif == true)
                    {
                        SimbolAktif((*M), S);
                	}
                }
                break;
            }
        case 'x':
            {
                if (ordinat <= 15)
                {
                    if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
                    ordinat += 4;
                    gotoxy(absis,ordinat);
                    if(!aktif)
                    {
                    	PointerAktif();
                    }
                    if (aktif == true)
                    {
                        SimbolAktif((*M), S);
                	}
                }
                break;
            }
        case 'c':
            {
                if ((absis <= 44)&&(ordinat <= 15))
                {
                    if(!aktif)
						{
							HapusSimbol(absis,ordinat);
						}
                    absis += 8;
                    ordinat += 4;
                    gotoxy(absis,ordinat);
                    if(!aktif)
                    {
                    	PointerAktif();
                    }
                    if (aktif == true)
                    {
                        SimbolAktif((*M), S);
                	}
                }
                break;
            }
		case 'r':
            {
            	if (aktif == false)
            	{
		        	Rotate(M);
		        	gotoxy(61,12);
		        	printf("                ");
		        	gotoxy(61,12);
		        	printf("%s",kata);
		        	gotoxy(61,14);
		        	printf("                ");
		        	BuatBoard();
		        	IsiBoard(M);
					initBoard();
            	}
            	break;
            }            
        case 's':
            {
                simpanabsis = absis;
                simpanordinat = ordinat;
                if (aktif == false)
                {
                    SimbolAktif((*M), S);
                    aktif = true;
                }
                else
                {
                    EndWord(S, (kata), endK);
					//Score(S, &score, &kata);
					//printf("score: %d", score);
					PointerAktif();
				}
                break;
            }
        default:
        	{
                /*gotoxy(30,22);
                printf("\r                                                ");
                gotoxy(30,22);
            	printf("doesn't recognized the input");*/
            	break;
            }
	}
}

void SimbolAktif(MATRIKS M, Stack *S)
{
    // Kamus Lokal
    infotype A;

    // Algoritma
    if (discovered[((ordinat+1)/4-1)][((((absis+8)/4)-7)/2)])
    {

        while ((GetAbsis(Pos(Infotop(*S))) != absis) || (GetOrdinat(Pos(Infotop(*S))) != ordinat))
        {
            HapusSimbol(GetAbsis(Pos(Infotop(*S))), GetOrdinat(Pos(Infotop(*S))));
            discovered[((GetOrdinat(Pos(Infotop(*S)))+1)/4-1)][((((GetAbsis(Pos(Infotop(*S)))+8)/4)-7)/2)] = false;
            Pop(S, &A);
        }
        absis = GetAbsis(Pos(Infotop(*S)));
        ordinat = GetOrdinat(Pos(Infotop(*S)));
        gotoxy(absis, ordinat);
    }
    else
    {
        gotoxy(absis-1,ordinat);
        Bold();
        printf("%s[", KRED);

        gotoxy(absis+1,ordinat);
        printf("]%s", KNRM);
        UnBold();
        gotoxy(absis,ordinat);

        SetAbsis(&Pos(A), absis);
        SetOrdinat(&Pos(A), ordinat);
        Info(A) = GetElmt(M, ((ordinat+1)/4-1), ((((absis+8)/4)-7)/2));

        Push(S, A);
        //gotoxy(30, 24 + count);
        //printf("%d %d %c", PosX(Infotop(*S)), PosY(Infotop(*S)), Info(Infotop(*S)));

        gotoxy(absis,ordinat);
        discovered[((ordinat+1)/4-1)][((((absis+8)/4)-7)/2)] = true;
        count++;
    }
}

void PointerAktif()
{
	gotoxy(absis-1,ordinat);
    printf("[");

    gotoxy(absis+1,ordinat);
    printf("]");
    gotoxy(absis,ordinat);
}

void HapusSimbol(int x, int y)
{
    // Kamus Lokal

    // Algoritma
    gotoxy(x-1,y);
    printf(" ");
    gotoxy(x+1,y);
    printf(" ");
    gotoxy(x,y);
}

void EndWord(Stack *S, char kata[18], boolean *endK)
{
    // Aamus Lokal
    infotype A;
    int i, j;

    // Algoritma
    i = TOP(*S);
    aktif = false;
    while (IsEmpty(*S) == false)
    {
        Pop(S, &A);
        (kata)[TOP(*S)] = Info(A);
        HapusSimbol(GetAbsis(Pos(A)), GetOrdinat(Pos(A)));
    }
    (kata)[i] = '\0';

    absis = GetAbsis(Pos(A));
    ordinat = GetOrdinat(Pos(A));
    for (i=1;i<=4;i++) {
        for (j=1;j<=4;j++) {
            discovered[i][j] = false;
        }
    }

    gotoxy(30,22);
    printf("                           ");
    gotoxy(30,22);

    (*endK) = true;
}

/* ADT Set */
void addFoundWord ( SetStr *E, char word[18] ) {
	SetSetEl(E, NbElmt(*E), word);
	SetNeff(E, (NbElmt(*E) + 1));
	//TulisIsiTab(E);
}

boolean isFound ( SetStr E, char word[18] ) {
	int i, n;
	ElType *kata;
	boolean found;
	
	if ( NbElmt(E) != 0 ) {
		for ( i=0; i<NbElmt(E); i++ ) {

			n = 0;
			found = true;
			while ( ( (E.TI[i][n] != '\0' ) || (word[n] != '\0') ) && ( found ) ) {
				if (E.TI[i][n] != word[n]) {
					found = false;
				}
				n++;
			}

			if ( found == true ) {
				break;
			}
		} 
	} else {
		found = false;
	}

	return found;
}


/* End of ADT Set */

int Map (char key)
{
	StrMap *sm;
	char buf[255];
	int result;
	int val;

	sm = sm_new(30);
	if (sm != NULL) {
	sm_put(sm, "E", "1");
	sm_put(sm, "A", "11");
	sm_put(sm, "I", "11");
	sm_put(sm, "N", "11");
	sm_put(sm, "O", "11");
	sm_put(sm, "R", "11");
	sm_put(sm, "S", "11");
	sm_put(sm, "T", "11");
	sm_put(sm, "C", "111");
	sm_put(sm, "D", "111");
	sm_put(sm, "L", "111");
	sm_put(sm, "G", "1111");
	sm_put(sm, "H", "1111");
	sm_put(sm, "M", "1111");
	sm_put(sm, "P", "1111");
	sm_put(sm, "U", "1111");
	sm_put(sm, "B", "11111");
	sm_put(sm, "F", "11111");
	sm_put(sm, "Y", "11111");
	sm_put(sm, "K", "111111");
	sm_put(sm, "V", "111111");
	sm_put(sm, "W", "111111");
	sm_put(sm, "Z", "11111111");
	sm_put(sm, "X", "111111111");
	sm_put(sm, "Q", "1111111111");
	sm_put(sm, "J", "1111111111");

	result = sm_get(sm, &key, buf, sizeof(buf));
	if (result != 0) {
	val = strlen(buf);
	return val;
	}
	}
}
 
int Score(char kata[18])
{
	int sum=0;
	int count=0;
	int score=0;
	//boolean endK;
	//EndWord(S,(*kata),&endK);
	do {
		sum += Map((kata)[count]);
		count++;
	} while (kata[count] != '\0');
	(score) = sum*(count);
	if (count>=10) {
		score += 100;
	}

	return score; 
}

void Rotate (MATRIKS *M)
{
	// Kamus Lokal
	
	// Algoritma
	Transpose (M);
	FlipVertical(M);
}
